File: data-structures/python/doubly_linked_list.py
from typing import TypeVar

T = TypeVar("T")


class Node:
    def __init__(self, value: T):
        self.value = value
        self.next = None
        self.previous = None


class DoublyLinkedList:
    _head: Node = None
    _tail: Node = _head
    _length: int = 0

    def __len__(self):
        return self._length

    def size(self):
        return self._length

    def __iter__(self):
        self._current = self._head
        self._index = -1
        return self

    def __next__(self):
        if self._index < self._length - 1:
            data = self._current.value
            self._current = self._current.next
            self._index += 1
            return data
        else:
            raise StopIteration

    def append(self, value: T):
        newNode = Node(value)
        if self._head == None:
            self._head = newNode
            self._tail = self._head
        else:
            self._tail.next = newNode
            newNode.previous = self._tail
            self._tail = self._tail.next
        self._length += 1

    def pop(self):
        self._tail = self._tail.previous
        self._tail.next = None

    def remove(self, index):
        if index == 0:
            self._head = self._head.next
            self._head.previous = None
        else:
            current = self._head
            for _ in range(0, index - 1):
                current = current.next
            current.next = current.next.next
        self._length -= 1

    def reverse(self):
        current = self._head
        temp = current.next
        while temp is not None:
            current.previous, current.next = current.next, current.previous
            current = temp
            temp = current.next
        current.previous, current.next = current.next, current.previous
        self._tail.previous = None
        self._head, self._tail = self._tail, self._head

    def indexOf(self, value: T):
        current = self._head
        index = 0
        if current.value == value:
            return index
        else:
            while current.next is not None:
                if current.value == value:
                    return index
                current = current.next
                index += 1
            return index if self._tail.value == value else -1

    def toString(self):
        current = self._head
        result = "["
        while current.next is not None:
            result += f"{current.value}, "
            current = current.next
        result += f"{current.value}]"
        print(result)

File: data-structures/typescript/doubly-linked-list.ts
type LNode<T> = {
    value: T,
    next?: LNode<T>,
    prev?: LNode<T>
}
class DoublyLinkedList<T>{
    public length: number;
    private head?: LNode<T>;
    private tail?: LNode<T>;
    constructor(){
        this.length = 0;
        this.head = undefined;
        this.tail = undefined;
    }
    prepend(item: T) :void {
        const node = {value: item} as LNode<T>;
        this.length++;
        if(!this.head){
            this.head = this.tail = node;
            return;
        }
        node.next= this.head;
        this.head.prev = node;
        this.head = node;
    }
    insertAt(item: T, idx: number): void {
        if (idx > this.length){
            throw new Error("oh no");
        } else if  (idx === this.length){
            this.append(item);
            return;
        } else if  (idx === 0){
            this.prepend(item);
            return;
        }
        this.length++;
        const curr = this.getAt(idx) as LNode<T>;
        const node = {value: item} as LNode<T>;
        node.next = curr;
        node.prev = curr.prev;
        curr.prev = node;
        if (node.prev){
            node.prev.next = node;
        }
    }
    append(item:T):void{
        const node = {value: item} as LNode<T>;
        this.length++;
        if (!this.tail){
            this.head = this.tail = node;
            return;
        }
        this.tail.next = node;
        node.prev = this.tail;
        this.tail = node;

    }
    remove(item:T): T| undefined{
       let curr = this.head;
       for (let i = 0; curr && i < this.length; i++) {
        if (curr.value === item) break;
        curr = curr.next;
       }
       if (!curr){
        return undefined;
       }
       return this.removeNode(curr);
    }
    get(idx:number) :T| undefined{
        return this.getAt(idx)?.value;
    }
    removeAt(idx:number): T|undefined{
        const node = this.getAt(idx);
        if (!node){
            return undefined;
        }
        return this.removeNode(node);
    }
    private removeNode(node: LNode<T>): T | undefined {
       this.length--;
       if (this.length === 0){
        const out = this.head?.value;
        this.head = this.tail = undefined;
        return out;
       }
       if( node.prev){
        node.prev.next = node.next;
       }
       if (node.next){
        node.next.prev = node.prev;
       }
       if (node === this.head){
        this.head = node.next; 
       }
       if (node === this.tail){
        this.tail = node.prev;
       }
       node.prev = node.next = undefined;
    return node.value; 
    }
    private getAt(idx: number): LNode<T> | undefined{
        let curr = this.head;
        for (let i =0; curr && i < idx; i++){
            curr = curr.next;
        }
        return curr;
    }
}